### DISCUSSION TITLE
How to specify which encoding to use when fetching/inserting data?

### DISCUSSION THREAD

**Discussion by FredericoCoelhoNunes** [sqlalchemy/sqlalchemy#6868] [Usage Questions]
Hi everyone,

I'm running into a bunch of issues when dealing with non-UTF-8 data during a database migration, and would really appreciate some help.

The issues are the following:

* when select a CHAR column from my Oracle database, SQLAlchemy correctly decodes it using latin-1 (I confirmed by manually fetching the bytes from the column, decoding them using latin-1, and then comparing the decoded strings with the strings that SQLAlchemy selected). However, when inserting them into my second database (MariaDB), SQlAlchemy decides to encode the strings as UTF-8, even though the column in the MAriaDB table clearly specifies the latin1 charset. How can I tell SQLALchemy to use latin-1 for this column?
* I tried to specify `?charset=latin1` in my MariaDB connection string, even though this isn't totally correct (I only want latin1 for this particular column). However, I wanted to test if this fixed the issue. In this case, I also got a weird error: `UnicodeDecodeError: 'charmap' codec can't encode character \x96` from `File "usr/lib64/python3.9/encodings/cp1252.py", line 12, in encode`. This seems to indicate that SQLAlchemy is using the WIndows-1252 encoding, even though I clearly specified it should use latin-1. How can I change this?

Thanks for reading, and if you have any insight please let me know! Cheers.

**Comment by zzzeek** [comment]
a few misconceptions:

* assuming you are on python 3, SQLAlchemy doesn't do any encoding or decoding.  it passes Python strings to the DBAPI drivers, which are unicode.  the drivers do all the bytestring stuff
* AFAIK, the encoding that is set up for the connection overall, like MySQL `?charset=<charset>`, is the only level of granularity you get for sending Python strings (unicode objects) through.   so any "this should be latin-1" has to be passed as bytes from the outside.
* it is *very unusual* to have a database schema / connection / etc where everything is utf-8, but just *one* column is latin-1.   I can't think of a technically valid reason to have your schema this way, as IIUC utf-8 can encompass any set of glyphs that latin-1 can.  if you need a latin-1 string to pass out to the real world you can get back the Python unicode object and then encode into latin-1.    But let's assume you want to keep doing this
* I believe that for most/all the MySQL drivers, if you do the encode into latin-1 yourself, you can send that right through and the drivers should let it right in, so try that.
* ultimately, if this is proving to put up continued resistance, you need to make a python script with just the MySQL driver you are using and get it to make a connection and push out a latin-1 bytestring.  if you can make the driver do it with a plain python script I can show you how to get SQLAlchemy to pass it through.

**Reply by FredericoCoelhoNunes** [reply]
Thanks a lot for the reply!
I have in the meanwhile been able to solve the problem - PyMySQL has the following code:

```
if charset == 'latin1':
    encoding = 'cp1252'
```
Which means it is using the Windows-1252 encoding even when latin-1 is requested. Now, I thought this was OK because from my understanding, Windows-1252 is a superset of ISO-8859-1 which I thought means that everything that can be encoded in ISO-8859-1 can also be encoded by Windows-1252. This is wrong, which can be seen with a minimal reproducible example:

```
x = '\x96'
x.encode('latin-1')
x.encode('cp1252')  # throws an error
```
So what was happening was:

* SQL Alchemy selects the column from the database, and some rows end up having Unicode character \x96 (the Python strings (not the bytes!) are like `"hello world \x96"`, where `my_str[-1] = 'x96'`. I have tried checking the name of this unicode character but Python throws an error, somehow the character has no name. Not exactly sure what this means).
* then PyMySQL tries to encode this '\x96' character with Windows-1252, which fails.

Problem was fixed by editing the PyMySQL source code and removing the latin-1 -> cp1252 conversion.

Cheers

**Reply by CaselIT** [reply]
> Problem was fixed by editing the PyMySQL source code and removing the latin-1 -> cp1252 conversion.

you may want to report this issues to pymysql, since this fix does not seem very portable :)

**Reply by FredericoCoelhoNunes** [reply]
I have already reported it @CaselIT ! I'd say it's 99% some misconception on my end, but maybe it will be helpful somehow. Cheers
